clients/src/main/java/org/apache/kafka/clients/ClusterConnectionStates.java

```diff
@@ -101,17 +108,51 @@ public boolean isConnected(String id) {
     public void connected(String id) {
         NodeConnectionState nodeState = nodeState(id);
         nodeState.state = ConnectionState.CONNECTED;
+        resetReconnectBackoff(nodeState);
     }
 
     /**
      * Enter the disconnected state for the given node
+     * and calculate backoff delay for next connection attempt
+     *
      * @param id The connection we have disconnected
      * @param now The current time
      */
     public void disconnected(String id, long now) {
         NodeConnectionState nodeState = nodeState(id);
         nodeState.state = ConnectionState.DISCONNECTED;
         nodeState.lastConnectAttemptMs = now;
+        updateReconnectBackoff(nodeState);
+    }
+
+    /**
+     * Resets the failure count for a node and sets the reconnect backoff to the base
+     * value configured via reconnect.backoff.ms
+     *
+     * @param nodeState The node state object to update
+     */
+    public void resetReconnectBackoff(NodeConnectionState nodeState) {
+        nodeState.failedAttempts = 0;
+        nodeState.reconnectBackoffMs = this.reconnectBackoffMs;
+    }
+
+    /**
+     * Update the node reconnect backoff exponentially, but with a randomly selected value
+     * in the range [base, min(max, base * 2**failures)], where base is configured via
+     * reconnect.backoff.ms and max is configured via reconnect.backof.max
+     *
+     * @param nodeState The node state object to update
+     */
+    public void updateReconnectBackoff(NodeConnectionState nodeState) {
+        if (this.reconnectBackoffMax > this.reconnectBackoffMs) {
+            nodeState.failedAttempts += 1;
+            double backoffExp = (double) Math.min(nodeState.failedAttempts - 1, this.reconnectBackoffMaxExp);
+            double backoffFactor = (double) Math.pow(this.reconnectBackoffExpBase, backoffExp);
+            long reconnectBackoffMs = (long) (Math.max(this.reconnectBackoffMs, 1) * backoffFactor);
+            // Actual backoff is chosen randomly in the exponentially-increasing range, which should make
+            // connection attempts during broker failure more uniformly distributed.
+            reconnectBackoffMs = ThreadLocalRandom.current().nextLong(this.reconnectBackoffMs, nodeState.reconnectBackoffMs);
```

Comment:
This is just setting the variable declared right before it. Presumably you mean to set `nodeState.reconnectBackoffMs`?